apps/passport-server/src/services/e2eeService.ts

```diff
@@ -52,21 +86,66 @@ export class E2EEService {
   ): Promise<void> {
     logger(`[E2EE] Saving ${request.blobKey}`);
 
-    await insertEncryptedStorage(
-      this.context.dbPool,
-      request.blobKey,
-      request.encryptedBlob
-    );
+    if (!request.blobKey || !request.encryptedBlob) {
+      throw new PCDHTTPError(400, "Missing request fields");
+    }
 
-    res.json(undefined satisfies UploadEncryptedStorageResponseValue);
+    let resultRevision = undefined;
+    if (request.knownRevision === undefined) {
+      resultRevision = await setEncryptedStorage(
+        this.context.dbPool,
+        request.blobKey,
+        request.encryptedBlob
+      );
+    } else {
+      const updateResult = await updateEncryptedStorage(
+        this.context.dbPool,
+        request.blobKey,
+        request.encryptedBlob,
+        request.knownRevision
+      );
+      resultRevision = this.checkUpdateResult(
+        request.blobKey,
+        request.knownRevision,
+        updateResult
+      );
+    }
+
+    res.json({
+      revision: resultRevision
+    } satisfies UploadEncryptedStorageResponseValue);
+  }
+
+  private checkRekeyResult(
```

Comment:
this function doesn't use `blobKey`.

Also it does more than just `check` - it returns something via the response value which is kinda sketchy to me.

Reply:
It uses the blobKey in an error message, though probably it shouldn't since that's sensitive data.  I'll remove it.

As for the name, it's the same pattern as checkQueryParam, which you wrote.  Do you think a different name would be clearer?

Reply:
Actually it looks like blobKey is currently all over the logs and error messages, so I'm not going to mess with it.  We can rethink later whether it should be obfuscated.

Reply:
there's a difference between `checkQueryParam` and this one - this check has a side effect which is that it writes to `res`. i've found this to be a common source of bugs - passing res around and conditionally writing to it. my preference is to figure out what to write and the write, in a sort of 2 step process. one more level of acceptable complexity is writing in multiple places in one function, and carefully using early returns. another thing we sometimes do is validate a request (as in the various routes files), and then pass `res` into a handler, and forget about it in the router, completely delegating returning a response via `res` to the handler.

all that being said it's not a huge deal if you leave it in this way, just something i've picked up over time.

Reply:
Ah, I see your point now.  Mutable arguments are an antipattern I prefer to avoid in general.  In this case it arose because of `res` which is morally a return value, but handled mutably by the framework here.  The first function like this which I wrote always threw on errors, but this one had to manipulate the result instead.  That messes with the cleanliness of the "continue only on success" pattern I was going for.

Reply:
I think the "right" answer might be to implement the TODO so I can return custom JSON in a PCDHTTPError.  I'd rather not further mess with the framework today, though, so I'll do something clearer locally instead.